Add async routing backend with first level filtering

danialmalik opened the pull request:
### PR description:

- Add async routing backend
This backend will process the events first (if there are any processors) and then pass the event to configured backends to be processed asynchronously

- Add a Regular Expression based filter that will be used to first filter the events before sending to the backends.
    - Individual filters will be created for each backend
    - Filters are of two types. Blacklist and whitelist
    - Do not let the event pass through if the backend has no configured filter settings available
    - ~~If there is no such filter for a backend, let the event pass through.~~
    - ~~Use the latest modified and enabled filter object for a backend~~
    - ~~These filters are configurable from Admin panel~~


P.S. Following tasks are in progress, but I’ll create a new PR for these:
- IMS Caliper transformation of events
- XAPI transformation of events
- Router for delivery of the events.


### Update
As per new changes, the following settings are required to be placed in edx-platform

```
EVENT_TRACKING_BACKENDS = {
    'caliper': {
        'ENGINE':  'eventtracking.backends.async_routing.AsyncRoutingBackend',
        'OPTIONS': {
            'backend_name': 'caliper',
            'processors': [
                {
                    'ENGINE': 'eventtracking.processors.regex_filter.RegexFilter',
                    'OPTIONS':{
                        'filter_type': 'allowlist',
                        'regular_expressions': [
                            'edx.course.enrollment.activated',
                            'edx.course.enrollment.deactivated',
                        ]
                    }
                }
            ],
            'backends': {
                'caliper': {
                    'ENGINE': 'edx_analytics_transformers.backends.caliper.CaliperBackend',
                    'OPTIONS': {}
                    }
                }
            },
        },
    },
    'xapi': {
        'ENGINE':  'eventtracking.backends.async_routing.AsyncRoutingBackend',
        'OPTIONS': {
            'backend_name': 'xapi',
            'processors': [
                {
                    'ENGINE': 'eventtracking.processors.regex_filter.RegexFilter',
                    'OPTIONS':{
                        'filter_type': 'allowlist',
                        'regular_expressions': [
                            'edx.course.enrollment.activated',
                            'edx.course.enrollment.deactivated',
                        ]
                    }
                }
            ],
            'backends': {
                'xapi': {
                    'ENGINE': 'edx_analytics_transformers.backends.xapi.XApiBackend',
                    'OPTIONS': {}
                    }
                },
            }
        }
    },
    'tracking_logs': {
    ...
    }
}
```

edx-webhook commented:
Thanks for the pull request, @danialmalik! I've created [OSPR-4811](https://openedx.atlassian.net/browse/OSPR-4811) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.


<!-- jenkins ok to test -->

natabene commented:
@danialmalik Thank you for your contribution. Please let me know once it is ready for our review.

danialmalik commented:
@natabene yes please, line it up for the review.

nasthagiri commented:
Glad to see this coming in!

**Question:** Can we put all of the xAPI and Caliper related code in a separate repository?  This allows us to keep this repo with the core plugin interface for event tracking - and have concrete backends separate from the core.

Assuming you can do the above, you can create a new PR that includes only those changes that need to be part of the core API.  All other functionality can be in a separate repo.

ziafazal commented:
@nasthagiri You mean we should move Caliper and xAPI related code into separate repo and make it a new django app i.e. `eventtracking_backends` which depend on core event tracking app(this repo) and both of these apps can be plugged into edx-platform or any other service to emit and rout events to relevant backends. 
![Event tracking (1)](https://user-images.githubusercontent.com/434995/87444689-8a486c80-c610-11ea-9e96-7040c2afd2be.png)

nasthagiri commented:
Right.  The new repo though wouldn't be for all `eventtracking backends` - only for xAPI/Caliper (for real-time standards-based learning events).

Cc: @brianhw @mulby

nasthagiri commented:
Also, Cc @feanil and @ormsbee as we think about [scaling inter-service eventing](https://github.com/edx/open-edx-proposals/pull/136) and how these standards-based eventing backends will integrate with our new upcoming infrastructure.

macdiesel commented on eventtracking/django/tests/test_models.py:
```diff
@@ -0,0 +1,113 @@
+        )
+
+    @ddt.data(
+        ('whitelist', 'sentinel.* \n not_matching \n .*', True),
```
We are trying to change this language across our code base from whitelist/blacklist to allowlist/blocklist. Could we make that change here as well?

macdiesel commented on eventtracking/django/models.py:
```diff
@@ -0,0 +1,174 @@
+    valid_regex_expressions = []
+    for exp in expressions_list:
+        try:
+            re.compile(exp)
```
Tracing through this code I think that compile here is being called each time an event is fired then the list of filters is iterated to find a match. Because these are compiled each time this code would not gain the benefits of the compile caching. Given that the velocity in which this code would be executed using the `is_string_a_match` method I would suggest refactoring this code. I would suggest improved caching of these regexes. 

Also it seems like these may be compiled twice for each iteration of the filter? Once here to validate them and then once to compile them for checking. Code path looks like is_string_a_match -> _clean_expressions (compiled again on line 128) -> _validate_expressions (compiled first here on line 50 to validate).

danialmalik commented:
@nasthagiri we have separated the Caliper related code from the core.

@brianhw we have refactored the code to better utilize the caching, please re-review the PR.

ziafazal commented:
@nasthagiri @macdiesel could you please have another look?

macdiesel commented on .gitignore:
```diff
@@ -34,3 +34,6 @@ diff_quality_pylint.html
 pylintrc_backup
+
+# VS code
+.vscode/
```
Remove. This should go in your local gitignore, not to the checked in one.

macdiesel commented on Makefile:
```diff
@@ -1,4 +1,5 @@
 export DJANGO_SETTINGS_MODULE=eventtracking.django.tests.settings
+export PYTHONPATH=$(PWD)
```
This strikes me as odd. I don't think you can assume that everyone's python path will be at the root of this directory. Also isn't . usually on the python path? I don't think this should be here.

danialmalik replied:
Removed this by changing `pytest` command to `python -m pytest`

macdiesel commented on eventtracking/backends/async_routing.py:
```diff
@@ -0,0 +1,52 @@
+        try:
+            processed_event = self.process_event(event)
+        except EventEmissionExit:
+            return
```
Log something here.

macdiesel commented on eventtracking/backends/async_routing.py:
```diff
@@ -0,0 +1,52 @@
+            return
+
+        try:
+            json_event = json.dumps(processed_event, cls=DateTimeJSONEncoder)
```
Lets combine this block and the block above.

macdiesel commented:
Docstrings need to be updated to match openedX standards: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#docstrings

macdiesel commented on eventtracking/django/django_tracker.py:
```diff
@@ -0,0 +1,175 @@
+
+        return result
+
+    def instantiate_from_dict(self, values):
```
Should this method be private to this class?

macdiesel commented on eventtracking/django/models.py:
```diff
@@ -0,0 +1,219 @@
+    BLOCKLIST = 'blocklist'
+    ALLOWLIST = 'allowlist'
+    FILTER_TYPES = (
+        (BLOCKLIST, 'Blocklist'),
```
Thank you.

macdiesel commented on eventtracking/django/models.py:
```diff
@@ -0,0 +1,219 @@
+        verbose_name='List of regular expressions',
+        help_text=('This should be a list of regular expressions, seperated '
+                   'by a newline, for the events to be filtered.'),
+        validators=(validate_regex_list, )
```
👍

macdiesel commented on eventtracking/django/models.py:
```diff
@@ -0,0 +1,219 @@
+        return cls._get_cached_filter(backend_name=backend_name)
+
+    @classmethod
+    def _get_cached_filter(cls, backend_name):
```
I don't think that all this regex caching is necessary, it's built in to the library. See here: https://stackoverflow.com/questions/12514157/how-does-pythons-regex-pattern-caching-work

That said you do need to properly call the regex and not re-create them every time you attempt to route a message.

danialmalik replied:
The cache that `re` maintains is a single process cache (and has a limit of 512 expressions). Using this `TieredCache` make sure that cache will be shared among processes.

macdiesel reviewed (dismissed):
I've gone through and looked a few things and left some comments. I did not have time to review the tests. 

FYI all of your string replacements need to be fixed up to meet the new python string formatting standard: https://realpython.com/python-formatted-output/.

danialmalik commented:
@macdiesel, we have implemented the feedback changes. Please have another look.

feanil commented on eventtracking/backends/async_routing.py:
```diff
@@ -0,0 +1,69 @@
+
+        for name, backend in six.iteritems(self.backends):
+            try:
+                json_backend = dumps(backend)
```
What is the `type` of the backend object here? It seems like it would be a python class from under [here](https://github.com/edx/event-tracking/tree/master/eventtracking/backends)?  It looks like you're using a new library to pass the class around?  That limits the kinds of classes that we can have and might potentially transmitting auth secrets for those backends as class attributes.  Why not just re-construct the backends from name in the worker?

feanil commented on requirements/base.in:
```diff
@@ -6,3 +6,7 @@ django
+celery
+django-model-utils
+edx-django-utils
+json-tricks
```
While this library is very clever it reaches into the innards of an object in order to figure out how to serialize it and that seems like a level of complexity that is not necessary to solve the problem it's solving.  I'd rather re-generate any python objects from input params than try to do this.

danialmalik commented on eventtracking/django/migrations/0001_initial.py:
```diff
@@ -0,0 +1,33 @@
+# Generated by Django 2.2.13 on 2020-07-26 09:33
+
+from django.db import migrations, models
```
@feanil we are following the [OEP-26](https://open-edx-proposals.readthedocs.io/en/latest/oep-0026-arch-realtime-events.html#filter-processor) for the development and its mentioned in the OEP to provide Admin UI for configurations.

If we don't want it to be this way, the OEP also needs to be modified.

feanil replied:
@danialmalik thanks for the pointer, I think we can update the OEP.  Since the self-service aspect is part of future work, I believe we can defer any models until we're ready to build out the APIs and UI to allow for self-service.  While the administration is being done by site administrators, I believe we should use django settings which are designed to be more easily managed by site admins.

feanil replied:
I've got a PR up on the OEP now: https://github.com/edx/open-edx-proposals/pull/156

ziafazal commented on eventtracking/tasks.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@task(name='eventtracking.tasks.async_send')
+def async_send(settings_name, json_event):
```
can rename it  to `send_event`?

danialmalik commented:
@feanil the django model and json_tricks have been removed now and the code is updated to get the filtering configurations from django settings. 

And if you are referring to the migration of `DjangoTracker` from `__init__.py` to `django_tracker.py`  in
>  A nice to have would be to move those formatting changes to a separate PR if it's not too hard.

this was only necessary when we were using the django model, since that is not the case now we can revert this change. Is that okay?

danialmalik commented:
Updating the PR description according to latest code

danialmalik commented on eventtracking/async_tracker.py:
```diff
@@ -0,0 +1,142 @@
+        Raises:
+            KeyError
+        """
+        filter_settings = getattr(settings, ASYNC_ROUTING_FILTERS_SETTINGS_NAME, {})[backend_name]
```
Removed `ASYNC_ROUTING_FILTERS_SETTINGS_NAME` setting requirement now

danialmalik commented on eventtracking/async_tracker.py:
```diff
@@ -0,0 +1,142 @@
+FilterConfig = namedtuple('FilterConfig', ['type', 'regular_expressions'])
+
+
+class AsyncTracker(DjangoTracker):
```
I've removed this tracker now. `DjangoTracker` will be used to initialize the tracker in celery tasks. The only limitation is that any tracking backend that we want to use asynchronously will have to be defined at root level of `EVENT_TRACKING_BACKENDS`.
Also, migrated the logic of the regular expression from `AsyncTracker` into `RegexFilter` processor class.

danialmalik commented on eventtracking/backends/async_routing.py:
```diff
@@ -0,0 +1,40 @@
+            )
+            return
+
+        send_event.delay(ASYNC_ROUTING_BACKENDS_SETTINGS_NAME, json_event)
```
Updated the celery task to use the existing `EVENT_TRACKING_BACKENDS` setting variable.

danialmalik commented:
@feanil please have another look at the updated code at your time of leisure. Thanks.

danialmalik commented:
@feanil I've updated the README and have added the `Asynchronous Routing` section. Could you please review the README updates.

I've also rebased and squashed the commits.

natabene commented:
@feanil Please re-visit once you are back.

feanil reviewed (approved):
Looks great! Please update the version here: https://github.com/edx/event-tracking/blob/master/setup.py#L53

You can set it to `0.4.0`

I can tag and release it and we've merged it.

danialmalik commented:
@feanil I've updated the version.

danialmalik commented:
@feanil I think the following change would be necessary once this PR is merged and the new version is released. Please have a look:
https://github.com/edx/edx-platform/pull/24954

feanil commented:
@danialmalik I missed that this was a breaking change, can you update the version to `1.0.0`   I'll add a note to the release as well.

danialmalik commented:
@feanil I've updated the version to `1.0.0`

edx-webhook commented:
@danialmalik 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/event-tracking&entry.867055334=https://github.com/edx/event-tracking/pull/56&entry.1484655318=https://github.com/danialmalik&entry.752974735=2020-07-09+08:12&entry.1917517419=2020-09-10+19:07&entry.2133058324=Yes) so we can improve your experience in the future.
